executor.go

```diff
@@ -714,14 +722,327 @@ func (e *executor) executeDifferenceShard(ctx context.Context, index string, c *
 	return other, nil
 }
 
+// RowIdentifiers is a return type for a list of
+// row ids or row keys. The names `Rows` and `Keys`
+// are meant to follow the same convention as the
+// Row query which returns `Columns` and `Keys`.
+// TODO: Rename this to something better. Anything.
+type RowIdentifiers struct {
+	Rows []uint64 `json:"rows"`
+	Keys []string `json:"keys,omitempty"`
+}
+
+// RowIDs is a query return type for just uint64 row ids.
+// It should only be used internally (since RowIdentifiers
+// is the external return type), but it is exported because
+// the proto package needs access to it.
+type RowIDs []uint64
+
+func (r RowIDs) merge(other RowIDs, limit int) RowIDs {
+	i, j := 0, 0
+	result := make(RowIDs, 0)
+	for i < len(r) && j < len(other) && len(result) < limit {
+		av, bv := r[i], other[j]
+		if av < bv {
+			result = append(result, av)
+			i++
+		} else if av > bv {
+			result = append(result, bv)
+			j++
+		} else {
+			result = append(result, bv)
+			i++
+			j++
+		}
+	}
+	for i < len(r) && len(result) < limit {
+		result = append(result, r[i])
+		i++
+	}
+	for j < len(other) && len(result) < limit {
+		result = append(result, other[j])
+		j++
+	}
+	return result
+}
+
+func (e *executor) executeGroupBy(ctx context.Context, index string, c *pql.Call, shards []uint64, opt *execOptions) ([]GroupCount, error) {
```

Comment:
Logically this makes sense given the limit/offset constraints, but the number of cluster round trips are large.   
Total Round Trips = # Rows calls + GroupBy 
I would like to package that into one trip and avoid the overhead of sending rows across the network

Reply:
yeah, we could do them async (so all at once), which would make it roughly 1 + GroupBy, but it still isn't ideal. The problem is that `Rows` calls which have a column argument should only be executed on the shard where that column lives, and so we *must* do those first before passing the resultant set of row ids to the GroupBy call on each shard. I think ideally, we'd do just the calls with `column` first and then pass the next ones on to each shard. In fact... I think we might need to do this for the code to be logically correct. I haven't yet written a test to verify this, but consider the following situation:

```
| field-row | shard 0 | shard 1 |
|-----------+---------+---------|
| a-0       | yes     | no      |
| a-1       | no      | yes     |
| b-0       | no      | yes     |
| b-1       | yes     | no      |
```

What does `GroupBy(Rows(field=a), Rows(field=b), limit=1)` return with the current logic? Remember that the lower limit of the `Rows` and `GroupBy` call is passed to each `Rows` query.

Reply:
Calling async doesn't limit the total amount of network traffic in my universe :).  I don't understand your table, what are the columns that are being intersected?

Reply:
Calling async doesn't limit the total amount of network traffic
true, but it affects the total latency which is arguably more important.

    I don't understand your table, what are the columns that are being intersected?

the leftmost is showing rows in various fields - the right two columns are showing whether or not that shard has data for that row.

The correct output for the query I gave would be something like `{Group: a.0-b.1, count: N}`, but I don't think it would return any results with the way the code is currently written. The rows passed to executeGroupByShard would just be [[0], [0]] and since no shard has row 0 in both a and b, no data would be returned.